docs_src/design/adr/security/0008-Secret-Creation-and-Distribution.md

```diff
@@ -0,0 +1,332 @@
+
+3. Dynamic injection of secret into container-scoped `tmpfs` volume
+
+   An example of this approach is [consul-template](https://github.com/hashicorp/consul-template).
+   This approach is useful when a secret is required to be in a configuration file
+   and cannot be passed via an environment variable
+   or directly consumed from a secret store.
+
+4. Distribution via _SECRETSLOC_
+
+   This option is the most widely supported secret distribution mechanism by container orchestrators.
+
+   EdgeX supports runtime environments such as standard Docker and snaps
+   that have no built-in secret management features.
+   
+   * Generic Docker does not have a built-in secrets mechanism.
+     Manual configuration of a _SECRETSLOC_ should utilize either
+     a host file file system path or
+     a Docker volume.
+   
+   * Snaps also do not have a built-in secrets mechanism.
+     The options for _SECRETSLOC_ are limited
+     to designated snap-writable directories.
+   
+   For comparison:
+
+   * Docker Swarm:
+   Swarm swarm mode is not officially supported by the EdgeX project.
+   Docker Swarm secrets are shared via the `/run/secrets` volume,
+   which is a Linux `tmpfs` volume created on the host and shared with the container.
+   For an example of Docker Swarm secrets, see the
+   [docker-compose secrets stanza](https://docs.docker.com/compose/compose-file/#secrets).
+   Secrets distributed in this manner become part of the RaftDB,
+   and thus it becomes necessary to enable swarm autolock mode,
+   which prevents the Raft database encryption key
+   from being stored plaintext on disk.
+   Swarm secrets have an additional limitation in that they are not
+   mutable at runtime.
+
+   * Kubernetes:
+   Kubernetes is not officially supported by the EdgeX project.
+   Kubernetes also supports the secrets volume approach,
+   though the secrets volume can be mounted anywhere in the container namespace.
+   For an example of Kubernetes secrets volumes, see the
+   [Kubernetes secrets documentation](https://kubernetes.io/docs/concepts/configuration/secret/).
+   Secrets distributed in this manner become part of the `etcd` database,
+   and thus it becomes necessary to specify a
+   [KMS provider for data encryption](https://kubernetes.io/docs/tasks/administer-cluster/kms-provider/)
+   to prevent `etcd` from storing plaintext versions of secrets.
+
+## Consequences
+
+As the existing implementation is not fully-compliant with this ADR,
+significant scope will be added to current and future EdgeX releases
+in order to bring the project into compliance.
+
+List of needed improvements:
+
+- PKI private keys
+  * All: Move to using Vault as system of origin for the PKI instead of the standalone `security-secrets-setup` utility.
+  * All: Cache the PKI for Consul and Vault on persistent disk; rotate occasionally.
+  * All: Investigate hardware protection of cached Consul and Vault PKI secret keys.  (Vault cannot unseal its own TLS certificate.)
```

Comment:
shouldn't this read "Kong" (not "Consul") and "Vault"?

Reply:
It was Consul because Vault depends on Consul.   Kong is started late enough in the flow that it is possible to reach into Vault and pull out a private key.

Reply:
** Note to self **  add treatment for a BYO kong cert.

Reply:
OK, but just to be clear this is cache PKI for future Consul usage, as AFIAK, there's no PKI material generated for Consul in Geneva, correct?

Reply:
> OK, but just to be clear this is cache PKI for future Consul usage, as AFIAK, there's no PKI material generated for Consul in Geneva, correct?

Correct. And not planning to do it in Hanoi either.

Reply:
BYO Kong added.